Add stock info to Products Report

psealock opened the pull request:
Fixes https://github.com/woocommerce/wc-admin/issues/936

Add stock quantity, status, and "low" status information to the Product and Product Detail Reports tables.

**Design: ** p6riRB-3dP-p2

### Also fixed here

`/report/variations` extended info was referencing the parent product, not the variation, for things like `permalink` and `image`. So that is fixed here along with relevant unit tests.

### Test

1. Edit a product's stock (_Edit Product > Inventory > Manage stock_).
1. _Products Report_
2. See your product in the table with an updated quantity and status.
3. Edit the "Low stock threshold" to a number above the quantity.
4. See your product in the table with a status of "Low" linked to the product edit page.
5. Repeat for variations of a product.
6. Note: if "manage stock" is turned off, a quantity won't be displayed.

psealock committed:
REST API: Add stock info to products and variations

Aljullu reviewed (commented):
It's testing well, but I noticed only the product name is displayed in the _Variations_ table:
![image](https://user-images.githubusercontent.com/3616980/49469765-fb396880-f7cd-11e8-9def-6cbaa6341bb2.png)

While before it was _Product name / Variation name_.

timmyc commented on client/analytics/report/products/table-variations.js:
```diff
@@ -123,11 +110,14 @@ export default class VariationsReportTable extends Component {
-					),
+					display:
+						'instock' === stock_status && stock_quantity && stock_quantity <= low_stock_amount ? (
+							<Link href={ editPostLink } type="wp-admin">
```
Kind of _feels_ like it should only show if the stock is low /cc @LevinMedia

psealock replied:
Designs have it only appear on when stock is low.

timmyc commented on client/analytics/report/products/table-variations.js:
```diff
@@ -82,15 +75,9 @@ export default class VariationsReportTable extends Component {
-			const name = this.getVariationName( row );
+			const { items_sold, gross_revenue, orders_count, extended_info, product_id } = row;
+			const { stock_status, stock_quantity, low_stock_amount } = extended_info;
+			const name = get( row, 'extended_info.name', '' ).replace( ' - ', ' / ' );
```
I did not know that was an eslint rule, TIL thanks for sharing that @Aljullu

psealock replied:
Fixed in https://github.com/woocommerce/wc-admin/pull/987/commits/74cce161fbe03683fe337d97d468474a74a7309d

timmyc commented on client/analytics/report/products/table-variations.js:
```diff
@@ -123,11 +110,14 @@ export default class VariationsReportTable extends Component {
-						</Link>
-					),
+					display:
+						'instock' === stock_status && stock_quantity && stock_quantity <= low_stock_amount ? (
```
Should we extract this logic into a util method since it is used twice in this PR alone? Also I'm thinking at some point we will need to suppress the low stock logic for sites that don't do inventory tracking. So if we had one util method that we based logic like this off of, it would be a singular spot to add tweaks like that.

psealock replied:
`isLowStock` has now been split into a util in https://github.com/woocommerce/wc-admin/pull/987/commits/0d4bd2e3f8b931c17a1eaed3bb0959f84910682c.

I placed it in a `/analytics/report/products/utils.js` file. Any opinion on that?

timmyc commented on includes/data-stores/class-wc-admin-reports-products-data-store.php:
```diff
@@ -157,6 +160,10 @@ protected function include_extended_info( &$products_data, $query_args ) {
 				}
+				// If there is no set low_stock_amount, use the one in user settings.
+				if ( '' === $extended_info['low_stock_amount'] ) {
+					$extended_info['low_stock_amount'] = absint( max( get_option( 'woocommerce_notify_low_stock_amount' ), 1 ) );
```
Is the fallback value of `1` what is used in Woo Core too?

psealock replied:
This is the same logic that will be introduced in https://github.com/woocommerce/wc-admin/pull/974. 

When the first of these PRs lands, it will be a good opportunity to consolidate into a util function.

timmyc commented on tests/reports/class-wc-tests-reports-products.php:
```diff
@@ -183,7 +183,11 @@ public function test_extended_info() {
 		$product = new WC_Product_Simple();
 		$product->set_name( 'Test Product' );
 		$product->set_regular_price( 25 );
+		$product->set_manage_stock( true );
```
Thanks for updating the tests here ⭐️

timmyc reviewed (commented):
Couple of quick questions, also going to tag @peterfabian for a review since this touches the REST endpoints.

peterfabian reviewed (commented):
Just some small notes.

psealock commented on includes/data-stores/class-wc-admin-reports-products-data-store.php:
```diff
@@ -60,6 +60,9 @@ class WC_Admin_Reports_Products_Data_Store extends WC_Admin_Reports_Data_Store i
 		'permalink',
+		'stock_status',
+		'stock_quantity',
+		'low_stock_amount',
```
Good call. I've done that in https://github.com/woocommerce/wc-admin/pull/987/commits/d205558cac27fb77a464381f60b18cdf6de4ecc5 and also updated `get_item_schema` for the Products Controller as well.

psealock commented on includes/data-stores/class-wc-admin-reports-variations-data-store.php:
```diff
@@ -125,20 +128,20 @@ protected function include_extended_info( &$products_data, $query_args ) {
-					}
-				}
-				$variations = $product->get_available_variations();
+				$variations          = $product->get_available_variations();
```
Nice, thank you. I added a check for the method in https://github.com/woocommerce/wc-admin/pull/987/commits/7aa7d191c3700ff11e3485bd8e512712172168bf

psealock commented:
Thanks for the reviews @Aljullu @timmyc @peterfabian !

I addressed the concerns and [fixed](https://github.com/woocommerce/wc-admin/pull/987/commits/3eb973bf619bc4214663b09d2ede10b9945c28c5) `$param_mapping` for variations that is needed due to recent changes.

>I noticed only the product name is displayed in the Variations table:

Hmm... I changed `name` to use whats coming from the backend instead of constructing it on the frontend. Maybe this change misses some cases? I created an issue to track that https://github.com/woocommerce/wc-admin/issues/1044.
